docs/zkapps/tutorials/02-private-inputs-hash-functions.mdx

```diff
@@ -79,35 +96,45 @@ Now we will add our initializer function.
 ```ts
 ...
  26
- 27   @method incrementSecret(secret: Field) {
+ 27   @method incrementSecret(salt: Field, secret: Field) {
  28     const x = this.x.get();
  29     this.x.assertEquals(x);
  30
- 31     Poseidon.hash([ secret ]).assertEquals(x);
- 32     this.x.set(Poseidon.hash([ secret.add(1) ]));
+ 31     Poseidon.hash([ salt, secret ]).assertEquals(x);
+ 32     this.x.set(Poseidon.hash([ salt, secret.add(1) ]));
  33   }
  34 }
  ```
 
- Mina uses the poseidon hash function, which has been optimized inside its proof system for fast performance. The poseidon hash function takes in an array of Fields, and returns a single Field as output.
+Mina uses the poseidon hash function, which has been optimized inside its proof system for fast performance. The poseidon hash function takes in an array of Fields, and returns a single Field as output.
+
+In this smart contract, we use both a secret number, and a second field, known as a "salt".
+
+In the `incrementSecret` function, we check that the hash of the salt and our secret is equal to the current state x. If this is the case, we add 1 to the secret and set x to the hash of the salt and the new secret. SnarkyJS creates a proof of this fact, which can be transmitted over the network to update the smart contract.
 
-We check that the hash of our secret is equal to the current state x. If this is the case, we add 1 to the secret and set x to the hash of that.
+Because zkApp smart contracts are run offchain, and only a proof is set over the network, both the salt and secret themselves remain private. Only the result, updating x to hash([ salt, secret + 1]) are revealed. Because the salt and secret can't be deduced from their hash, they remain private.
 
-When running this successfully, it just proves that the code was run with some input secret whose hash is x, and that the new x will be set to hash(secret + 1). However, the secret itself remains private, because it can't be deduced from it's hash.
+### But why the extra `salt` argument?
 
-### Main
+The extra `salt` argument is added to avoid a possible attack on our smart contract. If we were to just use `secret`, it would be vulnerable to discovery by an attacker. To do this, our attacker could try hashing likely secrets, and check if the hash matches the hash stored in the smart contract. If the hash were to match, then the attacker would have known they had discovered the secret. This is particularly concerning if the secret is likely to be in a particular range of the field, say between 1 and 10,000. In that case, with just 10,000 hashes, the attacker could discover our secret.
+
+What we can do to fix this, is add a second value, chosen uniformly over the field. This second value ensures that there are too many possibilities for an attacker to try for it to be possible to use this brute force method. This second value is commonly known as a "salt", and is the added field we've added to our smart contract.
```

Comment:
> chosen uniformly over the field

Can write this in a more accessible way?


> added field we've added to our smart contract

I could see this as being read as "stored as on-chain state" 

Maybe something along the lines of:

> What we can do to fix this, is add a second value as an input to our hash function, known as a "salt". The salt will be known only to us and is typically random for optimal security.  Using this second value as an additional input to our hash function ensures that an attacker cannot simply brute force attack our secret by generating hashes for likely values of it, because too many possibilities exist.